mne/beamformer/_dics.py

```diff
@@ -1,27 +1,581 @@
+    if reduce_rank and not (pick_ori == 'max-power' and inversion == 'matrix'):
+        raise NotImplementedError(
+            'The computation of spatial filters with rank reduction using '
+            'reduce_rank=True is only implemented with pick_ori=="max-power" '
+            'and inversion="matrix".'
+        )
+
+    # Inversion mode
+    if inversion not in ['single', 'matrix']:
+        raise ValueError("The inversion parameter should be either 'single' "
+                         "or 'matrix'.")
+
+    frequencies = [np.mean(freq_bin) for freq_bin in csd.frequencies]
+    n_freqs = len(frequencies)
+    n_orient = forward['sol']['ncol'] // forward['nsource']
+
+    # Determine how to normalize the leadfield
+    if normalize_fwd:
+        if inversion == 'single':
+            fwd_norm = 'dipole'
+        else:
+            fwd_norm = 'vertex'
+    else:
+        fwd_norm = None  # No normalization
+
+    picks = _setup_picks(info=info, forward=forward)
+    _, ch_names, proj, vertices, G = _prepare_beamformer_input(
+        info, forward, label, picks=picks, pick_ori=pick_ori,
+        fwd_norm=fwd_norm,
+    )
+    csd_picks = [csd.ch_names.index(ch) for ch in ch_names]
+
+    n_sources = G.shape[1] // n_orient
+
+    logger.info('Computing DICS spatial filters...')
+    Ws = []
+    for i, freq in enumerate(frequencies):
+        if n_freqs > 1:
+            logger.info('    computing DICS spatial filter at %sHz (%d/%d)' %
+                        (freq, i + 1, n_freqs))
+
+        Cm = csd.get_data(index=i)
+
+        if real_filter:
+            Cm = Cm.real
+
+        # Ensure the CSD is in the same order as the leadfield
+        Cm = Cm[csd_picks, :][:, csd_picks]
+
+        # Tikhonov regularization using reg parameter to control for
+        # trade-off between spatial resolution and noise sensitivity
+        # eq. 25 in Gross and Ioannides, 1999 Phys. Med. Biol. 44 2081
+        Cm_inv, _ = _reg_pinv(Cm, reg, rcond='auto')
+
+        if (pick_ori == 'max-power' and weight_norm == 'unit-noise-gain' and
+                inversion == 'matrix'):
+            Cm_inv_sq = Cm_inv.dot(Cm_inv)
+
+        # Compute spatial filters
+        W = np.dot(G.T, Cm_inv)
+
+        for k in range(n_sources):
+            Wk = W[n_orient * k: n_orient * k + n_orient]
+            Gk = G[:, n_orient * k: n_orient * k + n_orient]
+
+            # Compute power at the source
+            Ck = np.dot(Wk, Gk)
+
+            # Normalize the spatial filters
+            if Wk.ndim == 2 and len(Wk) > 1:
+                # Free source orientation
+                if inversion == 'single':
+                    # Invert for each dipole separately using plain division
+                    Wk /= np.diag(Ck)[:, np.newaxis]
+                elif inversion == 'matrix':
+                    # Invert for all dipoles simultaneously using matrix
+                    # inversion.
+                    Wk[:] = np.dot(linalg.pinv(Ck, 0.1), Wk)
```

Comment:
This is copied directly from the `make_lcmv` code, since the request was to have `make_dics`  be able to follow the `make_lcmv` approach exactly. This originates from some really old MNE-Python beamfomer code and survives to this day. This PR makes a small enhancement to the `lcmv._reg_pinv` function, namely the ability to specify `rcond='auto'`. The `'auto'` setting will estimate the rank of the matrix and attempts to set `rcond` accordingly. This change should be tested though, so maybe it's best left for the "merge lcmv and dics beamformer codepaths" project.

Reply:
Makes sense. At some point we should refactor this with the LCMV code. Hopefully someone is up for a big refactoring / unification PR during the next dev cycle... @britta-wstnr ? :)

Reply:
I overlooked this one before, sorry. Yes, @wmvanvliet is right, this code snippet is ancient and back then I could not trace any conversation or reviews on Github relating to it. It might be worth comparing this to FieldTrip's output (I only did this for the unit-noise-gain beamformer back then, but thus have the code to do so etc.), which might be a good starting point for fine tuning.

Refactoring: yes, let's do that!